### DISCUSSION TITLE
Really good game, and some features to improve

### DISCUSSION THREAD

**Discussion by bydariogamer** [beat-feet/beat-feet#52] [Ideas]
I found the game really creative. I am not actually used to Kotlin, so I will need a while to understand the code. As an issue: it takes too much to load the levels for the first time. Maybe you could try to have a comprised version of the platforms (or whatever way you use to check the generate the map and check the collisions) from the original sound files, instead of generating them when you select a game.
I found really difficult to play. The game is OK, the controls are not difficult, but the fact if is I found nearly impossible to walk too much in a level... I always die. I am not a actually good player, but there are situations in which you have two street lights and if you miss jumping into one there are too many obstacles until the floor is plain again amd you die for sure... Maybe a gaussian transformation to smooth the obstacles would be cool. And when collision is produced, I would suggest not to let player get damage again until it is not colliding.
Otherwise, the game is fantastic and I will try to improve my skills to end at leats a level.

**Comment by pserwylo** [comment]
Thanks so much for the kind words and honest feedback.

Regarding time to load: v0.2.0 comes with precompiled levels so the initial load is much faster. 

V0.2.0 also changes the way levels are generated a little to take into account not just overall sound, but low, mid, and high frequencies. This means many more obstacles. But don't fear! In this case more obstacles actually means some things are easier because there is less gaps, and less situations like the lights that you described. 

It is still frighteningly difficult though.

At this point, I think about it like flappy bird, or ghosts and ghouls, in that part of the allure of those games is  that it was freakishly hard to progress far, but still fun. Having said that, if it starts to take away from the fun then we should really reassess.

I like you're suggestion of something like a Gaussian bpurof obstacles, that is the kind of thinking that works well here (general purpose algorithmic solutions that would still work for custom MP3 files too). 

FDroid should build v0.2.0 soon, please let me know what you think when it becomes available!

Once that is built, I willb also tag v0.3.0 for a release soon after, which has level unlocking based on achievements.

**Comment by TheFlipside** [comment]
I also find the difficulty a little steep, it would help if you was able to collect items which replenish your health, maybe they could spawn when you go a certain distance without hitting anything, as reward on the way.

**Reply by pserwylo** [reply]
Ah, great call, this is something I've thought would he good too, especially if the items were also generated according to the music.

Furthermore, the next release (when FDroid build bot catches up) has an ever increasing rainbow following the character as your combo increases. Looks wonderful, but is not yet connected to any mechanic. I think that should fill up a rainbow bar of some sort, which can be used to... what?

I experimented with letting people "spend" it by triple jumping instead of just double jumping, but it was too hard to worry about it as a player. 

Now, I'm thinking when it fills up then perhaps we can give back health, either by showing a new button you can touch, or perhaps adding hearts to the screen that can be picked up by jumping into them. 

Any thoughts? (Not sure if I've explained very well).

P.s. thanks so much ;)

**Reply by TheFlipside** [reply]
That sounds good, the idea of a health item spawning

**Reply by bydariogamer** [reply]
I find the idea fantastic. What if the impossible obstacles are cut so they can be jumped and in the top of where it used to be there was a heart that heals you... let's say 10 Heal Points? For example:
![beatgame heart suggestion](https://user-images.githubusercontent.com/62303655/108981825-690f2a80-768d-11eb-8f48-33ffdb81a9a9.png)

**Reply by pserwylo** [reply]
That would be good. Kind of summarises my thoughts in #2, but it is not so much of an issue any more (will try to publish some pre-release versions of v0.2.0 and v0.3.0 tonight or tomorrow to showcase). 

I think this is a nice idea though, so how about:
* Fill up rainbow "power" by achieving combo's.
* At any point, you can "spend" it (e.g. by touching the rainbow bar) to cash it in.
* When cashed in, hearts spawn in front of you on the level.
* As per your screenshot, make them sit on top of various obstacles.
* If you wait for the bar to reach full power, you get more hearts spawning.
* However, if you really just need that little hit of health to keep you alive, you can trade in for a smaller amount.

Any thoughts? Alternatives? Concerns?

Once concern I have is making it obvious that the rainbow can be traded in for health. Although, after a quick play in Gimp, this looks okay to me (we can even explain very explicitly that saving up for longer will benefit you more):

![image](https://user-images.githubusercontent.com/248565/108984783-6ab80d00-76e4-11eb-9f2d-278c82ea727c.png)

**Reply by bydariogamer** [reply]
>     * Fill up rainbow "power" by achieving combo's.

Just like the punctuation in 0.1.0? It seems cool. I am not sure about what the rainbow power should be used for, but combos are always cool hehe

>     * At any point, you can "spend" it (e.g. by touching the rainbow bar) to cash it in.

Spending the combo looks good too. I would make a button somewhere else, because touching the screen that way makes you lose visibility (you would have to put your left hand over the character).
I am not sure about the idea of casting it at any time... Maybe if you had to wait until it is full, and then the button to trigger the power would be more challenging 🤔 

>     * When cashed in, hearts spawn in front of you on the level.
Also, the level could speed up and all the obstacles had 1/2 height. Another possibility is to add diamonds to get more points now. (The music could play at 2x speed during the rainbow-mode so the song doesn't mismatch the map).

I challenged a friend to bit my marks in 0.1.0 and I think it would be a cool additional option if there this kind of extra-points.

>     * As per your screenshot, make them sit on top of various obstacles.

Yea, it was an example of how the levels would be transformed. Actually I did it wrong because the character (what about giving per a name 👀) should have picked it.
By the way, I love that you use 'per' 😉.

>     * If you wait for the bar to reach full power, you get more hearts spawning.
>     * However, if you really just need that little hit of health to keep you alive, you can trade in for a smaller amount.

If you decide to make available to trigger the rainbow at any time it would be the better option for the player.

> Once concern I have is making it obvious that the rainbow can be traded in for health. Although, after a quick play in Gimp, this looks okay to me (we can even explain very explicitly that saving up for longer will benefit you more):

As I say above: it would be more challenging to have to wait to have it full for triggering it. But the other option is good too and your GIMP 'quick-edition' seems to be self-explaining.
There is a third option in which you can spend it at any time, and if you get to fill it, you can trigger an 'extra power' button.
My ideas seem to complicated to implement right now, so I would recommend to do as you wish, and don't lose yourself in a 'feature-spiral-of-dead'.

**Comment by toumbo** [comment]
Those lampposts are somewhat frustrating, also it'll help our jumps to be in coordination with the beat

**Comment by pserwylo** [comment]
Some interesting discussion at #59 regarding how to ease the difficulty:

> You could implement a shield that repairs max 20% but recharges by earning points

I think this might actually be a nicer mechanic for what to do with our ever increasing rainbow power achieved by making bigger combo's. I think it is better because it allows the same mechanism as the hearts discussed above (i.e. doing well in the level allows you to avoid pain), but it doesn't require the player to touch any more buttons, or to really interact in any othe rway (keeping the simplicity of just a single tap for jumping.

I'm going to create a prototype for this, which:
* Shows a rainbow bar in the corner near the hearts.
* Puts a shield icon next to the rainbow bar
* Uses visual feedback to let the user know that their shield is beign drained when they are hit. Depending on difficulty to implement, it could be:
** A flash and particles around the actual shield bar, including an animated drain as it drops, or
** Shield-like particles exploding from the player when they hit something, as well as a crest-shaped shield shown in front of the player when hit. Change to red particles and no crest shape when the shield is depleted.
* When the shield hits zero, health is drained
* When combo's are made, the shield is recharged.

**Reply by pserwylo** [reply]
Phew, so after a bit of investigating, I've gone ahead and added the shield idea. Wonderful suggestion!

https://github.com/beat-game/beat-game/releases/tag/v0.4.0 currently has a pre-release debug build that you can check out.

The shield takes damage at 5 times the normal rate (so empties quite quickly when getting hit) and recharges as you jump - recharging faster with bigger combos.

Overall, this makes the game much simpler (though we can tweak the difficulty by reducing the effectiveness of the shield - as I really like that it is quite difficult).

**Reply by TheFlipside** [reply]
This is great work, I installed the 0.4 version today and the game is much easier to play/beat now, in fact it was the first time i was ever able to finish a level without dying. This is exactly what we asked for, well done.

**Reply by pserwylo** [reply]
Glad to hear it is what you were after, and great to hear you finishing a level! I consider myself pretty good at this game, and before shields, I could only finish two levels. With this, I've been able to clear quite a few, so I may drop the effectiveness of the shields a bit in the future to keep a little bit of the chaos :)

**Comment by Glitchy-Tozier** [comment]
> What if the impossible obstacles were cut

I agree that there shouldn't be impossible obstacles... just for the record though: if this is the lamp in the beginning of the first level, it's not impossible. You can double-jump onto it.

**Reply by bydariogamer** [reply]
I didn't meant that lamp exactly, but in the last version I didn't found unjumpable obstacles.

**Reply by Glitchy-Tozier** [reply]
I see!  :)
